core/src/impl/Kokkos_HostSpace_deepcopy.cpp

```diff
@@ -0,0 +1,134 @@
+#include "Kokkos_HostSpace_deepcopy.hpp"
+
+namespace Kokkos {
+
+namespace Impl {
+
+/*
+ * TODO: change casting to whatever C++ likes (reinterpret_cast)
+ *       adjust logic with nthreads==1 to test if inside a parallel region
+ *       the reason for arranging the logic this way, is because
+ *       calling the omp_* API can incur some latency.
+ */
+
+void hostspace_parallel_deepcopy(void * dst, const void * src, size_t n) {
+  constexpr size_t page_sz_ = 4*1024;
+
+  // on KNL avoid having each thread requesting whole pages of memcpy
+  // on HSW, this may be equal to bytes_per_chunk
+  constexpr size_t bytes_per_chunk = page_sz_/2;
+
+  // make sure this is atleast 1 page, so that the calculation
+  // in the remainder below is valid unisgned arithmetic
+  if( n < 3*page_sz_)  {
+    std::memcpy( dst, src, n );
+    return;
+  }
+
+  // I can't use exec_space::concurrency here (or can I?)
+  // this may require specializing for each execSpace...
+  #ifdef _OPENMP
```

Comment:
If the callee passes on the execution space we can make the right decision here. Having _OPENMP defined doesn't necessarily mean you can use it here, trvially.

Reply:
I fixed this.